extensions/skins/conversation_v1/Conversation.js

```diff
@@ -75,227 +218,425 @@ oppia.directive('conversationSkin', [function() {
 
+      $scope.submitUserRating = function(ratingValue) {
+        ratingService.submitUserRating(ratingValue);
+      };
+      $scope.$on('ratingUpdated', function() {
+        $scope.userRating = ratingService.getUserRating();
+      });
+
+      $scope.$watch('currentProgressDotIndex', function(newValue) {
+        if (newValue !== null) {
+          _navigateToCard(newValue);
+        }
+      });
+
+      $scope.isWindowNarrow = function() {
+        return $(window).width() < 700;
+      };
+
+      $window.addEventListener('beforeunload', function(e) {
+        if (hasInteractedAtLeastOnce && !$scope.finished &&
+            !$scope.isInPreviewMode) {
+          oppiaPlayerService.registerMaybeLeaveEvent();
+          var confirmationMessage = (
+            'If you navigate away from this page, your progress on the ' +
+            'exploration will be lost.');
+          (e || $window.event).returnValue = confirmationMessage;
+          return confirmationMessage;
+        }
+      });
+
+      $scope.windowWidth = windowDimensionsService.getWidth();
       $window.onresize = function() {
         $scope.adjustPageHeight(false, null);
+        $scope.windowWidth = windowDimensionsService.getWidth();
+        _recomputeAndResetPanels();
+      };
+
+      $window.addEventListener('scroll', function() {
+        fadeDotsOnScroll();
+        fixSupplementOnScroll();
+      });
+
+      var fadeDotsOnScroll = function() {
+        var progressDots = $('.conversation-skin-progress-dots');
+        var progressDotsTop = progressDots.height();
+        var newOpacity = Math.max(
+          (progressDotsTop - $(window).scrollTop()) / progressDotsTop, 0);
+        progressDots.css({opacity: newOpacity});
+      };
+
+      var fixSupplementOnScroll = function() {
+        var supplementCard = $('md-card.conversation-skin-supplemental-card');
+        var topMargin = $('.navbar-container').height() - 20;
+        if ($(window).scrollTop() > topMargin) {
+          supplementCard.addClass('conversation-skin-supplemental-card-fixed');
+        } else {
+          supplementCard.removeClass('conversation-skin-supplemental-card-fixed');
+        }
+      };
+
+      $scope.canWindowFitTwoCards = function() {
+        return $scope.windowWidth >= $scope.TWO_CARD_THRESHOLD_PX;
+      };
+
+      $scope.initializePage();
+      ratingService.init(function(userRating) {
+        $scope.userRating = userRating;
+      });
+
+      oppiaPlayerService.getUserProfileImage().then(function(result) {
+        $scope.profilePicture = result;
+      });
+    }]
+  };
+}]);
+
+oppia.directive('answerFeedbackPair', [function() {
+  return {
+    restrict: 'E',
+    scope: {
+      answer: '&',
+      feedback: '&',
+      profilePicture: '&',
+      shortAnswer: '&'
+    },
+    templateUrl: 'components/answerFeedbackPair'
+  };
+}]);
+
+oppia.directive('progressDots', [function() {
+  return {
+    restrict: 'E',
+    scope: {
+      getNumDots: '&numDots',
```

Comment:
I wasn't aware you could do this. This would be super nice for many other variable bindings.